### ISSUE TITLE
[QUESTION] is it possible to fail a test in the QUnit.testDone callback?

### ISSUE THREAD

**Issue by williamsbdev** [jquery/qunit#817]
In the QUnit 2.0 series, the `assert` is no longer global (good thing!) and is passed into each test's callback. When I made this upgrade, my testDone callback is no longer failing tests like it had been (because I had been using the global assert). I had some code that would check to see that all requests were properly mocked and if there were too many or too few mocks, I would fail the test. I see my error messages but doing a `QUnit.assert.ok(false, "error message")` has no effect on the test. Can someone please advise how I would be able to fail the test that just run from inside the testDone callback?

Thanks!

**Comment by jzaefferer** [comment]
That's an interesting usecase, though one that the logging callbacks like `testDone` were never intended for. For this kind of assertion, I recommend using the module's `afterEach` callback instead. You may still need a bit of duplication to apply that check for all modules, but that's currently the best option. Let us know if that works for you.

**Comment by williamsbdev** [comment]
This will be in every module and so it is not desirable to put this into each `afterEach`. Would you like me to open a feature request issue?

**Comment by jzaefferer** [comment]
Assuming you abstract the necessary logic, you'd add 3-4 LOC per module. Also assuming one module per file, a reader of that file could quickly figure out where that shared setup is defined, compared to having it in just one place somewhere else.

That said, we've had a discussion about that feature before, even implemented it at some point, but dropped it before release. I can't remember that name of the ticket right now, but can look for it later...

**Comment by williamsbdev** [comment]
Correct me if I'm wrong but even in the `afterEach`, it is not possible to fail a test. So while I really don't want to add something to every module, I don't think it's even possible without wrapping the `test` function and doing my custom failure.

**Comment by gibson042** [comment]
http://jsbin.com/kotekikoti/1/edit?html,js,output

**Comment by jzaefferer** [comment]
So the answer to the question in the subject, can a `testDone` callback fail a test, is no, it cannot. The suggested workaround is to use module level callback (beforeEach, afterEach).

The attempt to implement global beforeEach/afterEach callbacks can be found in #633, while #681 discusses why we ended up removing them again. As I wrote there:

> it should be trivial to bring this back in whatever form we decide on later.

That's still a discussion we can have, though since this issue isn't about that, I'm going to close this for now.
